onnxruntime/core/providers/dml/DmlExecutionProvider/src/DmlRuntimeFusedGraphKernel.cpp

```diff
@@ -0,0 +1,374 @@
+                    isInputsUploadedByDmlEP.data(),
+                    isInputsUploadedByDmlEP.size(),
+                    m_isInitializerTransferable,
+                    m_partitionNodePropsMap,
+                    device.Get(),
+                    providerImpl,
+                    m_modelPath,
+                    m_subgraphNodePointers,
+                    m_subgraphInputs,
+                    m_subgraphOutputs);
+
+                m_outputShapes = graphDesc.outputShapes;
+
+                // Walk through each graph edge and mark used inputs
+                m_inputsUsed.resize(fusedNodeInputCount, false);
+                for (const DML_INPUT_GRAPH_EDGE_DESC& edge : graphDesc.inputEdges)
+                {
+                    m_inputsUsed[edge.GraphInputIndex] = true;
+                }
+
+                // Compile the operator
+                m_compiledExecutionPlanOperator = DmlGraphFusionHelper::TryCreateCompiledOperator(
+                    graphDesc,
+                    *m_indexedSubGraph,
+                    providerImpl);
+
+                TranslateAndCompileGraph(
+                    Info(),
+                    initializeResourceRefs,
+                    initInputBindings);
+            }
+
+            // Wrap tensors as required by Dml::IExecutionProvider::ExecuteOperator
+            OpKernelContextWrapper contextWrapper(
+                kernelContext,
+                Info().GetExecutionProvider(),
+                true,
+                nullptr);
+
+            ORT_THROW_IF_FAILED(m_provider->AddUAVBarrier());
+
+            // Get input resources for execution, excluding those which were specified as owned by DML and provided
+            // at initialization instead.
+            std::vector<ComPtr<IMLOperatorTensor>> inputTensors(kernelContext->InputCount());
+            std::vector<ID3D12Resource*> inputPtrs(kernelContext->InputCount());
+
+            for (int i = 0; i < kernelContext->InputCount(); ++i)
+            {
+                if (!m_inputsUsed[i])
+                {
+                    continue;
+                }
+
+                ORT_THROW_IF_FAILED(contextWrapper.GetInputTensor(i, inputTensors[i].GetAddressOf()));
+                inputPtrs[i] = m_provider->DecodeResource(MLOperatorTensor(inputTensors[i].Get()).GetDataInterface().Get());
+            }
+
+            auto aux = contextWrapper.GetOutputTensors(m_outputShapes);
+            ExecuteOperator(
+                m_compiledExecutionPlanOperator.Get(),
+                m_persistentResourceBinding ? &*m_persistentResourceBinding : nullptr,
+                inputPtrs,
+                aux);
+
+            ORT_THROW_IF_FAILED(m_provider->AddUAVBarrier());
+
+            // Queue references to objects which must be kept alive until resulting GPU work completes
+            m_winmlProvider->QueueReference(m_compiledExecutionPlanOperator.Get());
+            m_winmlProvider->QueueReference(m_persistentResourceAllocatorUnk.Get());
```

Comment:
>m_persistentResourceAllocatorUnk [](http://example.com/codeflow?start=44&length=32)

Should this simply call the EP allocator?  This can currently result in multiple allocations as well as multiple simultaneous allocations in the case of pipelined evaluations

Reply:
I forgot to remove those lines when I moved them to TranslateAndCompileGraph. Removing them now.

Reply:
The allocator alone won't know if it's safe to free this resource, so we need to keep it alive. We could have 2 successive calls to `Compute` with different sizes, and the second call could come before the first one is done executing on the GPU. In that scenario we would free the first `m_persistentResourceAllocatorUnk`, therefore signaling the allocator that we don't need it anymore, even though that's not the case.

Reply:
As long as both evaluations and the work to populate the persistent resource occurs on the same queue, this seems OK, if I'm not missing something.